Colab notebook (dataloader, downloading dataset)

Madhudorai opened the pull request.

Madhudorai committed:
Uploading colab notebook

iranroman commented:
```
- "directinput" and "micamp" are loaded from the audio folder.
- "exo" and "ego" are loaded from the video folder.
```
This statement is not necessary.

iranroman commented:
```
- slice_size (float): If specified (in seconds), samples are segmented into contiguous slices of this length.
                           The last slice, if shorter than slice_size, will be padded with zeros.
```

Change the name of this variable to `slice_dur`. Also state what the behavior is if `None`. I believe the default should not be `None` though. 

You should also have a variable to denote the overlap between slices. Do not assume that slices wont overlap.

iranroman commented:
```
- slice_range (tuple): Alternatively, a fixed (start, end) time window for all samples (in seconds).
                             Only one of slice_size and slice_range may be set.
```
Can you explain why this alternative is necessary?

```
        if slice_size and slice_range:
            raise ValueError("Cannot specify both slice_size and slice_range.")
```
Similarly, this may not be necessary.

iranroman commented:
```
        # Download and extract the dataset if it's not already present.
        if not os.path.exists(self.root_dir):
            print(f"{self.root_dir} not found. Downloading dataset...")
            zip_path = "dataset.zip"
            url = "https://zenodo.org/api/records/14963133/files-archive"
            urllib.request.urlretrieve(url, zip_path)
            print("Download complete. Extracting dataset...")
            with zipfile.ZipFile(zip_path, 'r') as zip_ref:
                zip_ref.extractall(self.root_dir)
            os.remove(zip_path)
            self._extract_nested_zip(self.root_dir)
            print("Dataset downloaded and extracted successfully.")
```

turn this into a method please. Also, please add a progress bar to show the download process. 
a potential long wait may be perplexing otherwise to users.

iranroman commented:
```
        self.players = self.available_players if players in ['all', ['all']] else players
        self.modalities = self.valid_modalities if modalities in ['all', ['all']] else modalities
        assert all(m in self.valid_modalities for m in self.modalities), \
            f"Modalities must be a subset of {self.valid_modalities}"
```
should you do a similar check here for the players?

iranroman commented:
```
            valid_contents = self.available_content[player]
            selected_contents = valid_contents if content_types in ['all', ['all']] else content_types
            for content in selected_contents:
                if content not in valid_contents:
                    continue
```

should you print a message here to warn the user about a content they want for a player but does not exist?

iranroman commented:
```
        # Define available players and content types.
        self.available_players = ['P1', 'P2', 'P3']
        self.available_content = {
            'P1': ['chords', 'scales', 'singlenotes', 'techniques'],
            'P2': ['chords', 'scales', 'singlenotes', 'techniques'],
            'P3': ['music']
        }
        self.valid_modalities = ['directinput', 'micamp', 'exo', 'ego']
```

make these variables uppercase to denote that they are GLOBAL definitions.

iranroman commented:
```
        # Build expanded_index for slicing. Use ceiling division so the last slice may be incomplete.
        if self.slice_size:
            self.expanded_index = []
            for i, sample_meta in enumerate(self.index):
                midi_path = self._get_midi_path(sample_meta)
                if os.path.exists(midi_path):
                    midi_obj = pretty_midi.PrettyMIDI(midi_path)
                    total_duration = midi_obj.get_end_time()
                    num_slices = int(np.ceil(total_duration / self.slice_size))
                    for s in range(num_slices):
                        start = s * self.slice_size
                        end = (s + 1) * self.slice_size  # May exceed total_duration; will be padded.
                        self.expanded_index.append((i, start, end))
        elif self.slice_range:
            self.expanded_index = [(i, self.slice_range[0], self.slice_range[1])
                                   for i in range(len(self.index))]
        else:
            # No slicing: entire sample is one slice.
            self.expanded_index = [(i, None, None) for i in range(len(self.index))]
```

This block of code I suggest you do it using vectorized operations. Look at librosa's audio slicing utilities.

Madhudorai commented:
> ```
> - slice_range (tuple): Alternatively, a fixed (start, end) time window for all samples (in seconds).
>                              Only one of slice_size and slice_range may be set.
> ```
> 
> Can you explain why this alternative is necessary?
> 
> ```
>         if slice_size and slice_range:
>             raise ValueError("Cannot specify both slice_size and slice_range.")
> ```
> 
> Similarly, this may not be necessary.

instead of segmenting modalities into slices of duration 5 second, (0-5,5-10, etc), we can instead new segments like (2-12) depending on what the user wants

Madhudorai commented:
> ```
> - slice_size (float): If specified (in seconds), samples are segmented into contiguous slices of this length.
>                            The last slice, if shorter than slice_size, will be padded with zeros.
> ```
> 
> Change the name of this variable to `slice_dur`. Also state what the behavior is if `None`. I believe the default should not be `None` though.
> 
> You should also have a variable to denote the overlap between slices. Do not assume that slices wont overlap.

if none- doesnt segment any of the audios- full length. okay will add a variable for overlap between slices

Madhudorai committed:
removing cell outputs

Madhudorai committed:
changes suggested by mr.iran

1. progress bar download
2. global variables
3. slice duration (not slice size)
4. added assertion, warning

Madhudorai commented:
> ```
> - "directinput" and "micamp" are loaded from the audio folder.
> - "exo" and "ego" are loaded from the video folder.
> ```
> 
> This statement is not necessary.

fixed in Commit f248e2b

Madhudorai commented:
> ```
>         # Download and extract the dataset if it's not already present.
>         if not os.path.exists(self.root_dir):
>             print(f"{self.root_dir} not found. Downloading dataset...")
>             zip_path = "dataset.zip"
>             url = "https://zenodo.org/api/records/14963133/files-archive"
>             urllib.request.urlretrieve(url, zip_path)
>             print("Download complete. Extracting dataset...")
>             with zipfile.ZipFile(zip_path, 'r') as zip_ref:
>                 zip_ref.extractall(self.root_dir)
>             os.remove(zip_path)
>             self._extract_nested_zip(self.root_dir)
>             print("Dataset downloaded and extracted successfully.")
> ```
> 
> turn this into a method please. Also, please add a progress bar to show the download process. a potential long wait may be perplexing otherwise to users.

fixed in Commit f248e2b

Madhudorai commented:
> ```
>         self.players = self.available_players if players in ['all', ['all']] else players
>         self.modalities = self.valid_modalities if modalities in ['all', ['all']] else modalities
>         assert all(m in self.valid_modalities for m in self.modalities), \
>             f"Modalities must be a subset of {self.valid_modalities}"
> ```
> 
> should you do a similar check here for the players?

fixed in Commit f248e2b

Madhudorai commented:
> ```
>             valid_contents = self.available_content[player]
>             selected_contents = valid_contents if content_types in ['all', ['all']] else content_types
>             for content in selected_contents:
>                 if content not in valid_contents:
>                     continue
> ```
> 
> should you print a message here to warn the user about a content they want for a player but does not exist?

fixed in Commit f248e2b

Madhudorai commented:
> ```
>         # Define available players and content types.
>         self.available_players = ['P1', 'P2', 'P3']
>         self.available_content = {
>             'P1': ['chords', 'scales', 'singlenotes', 'techniques'],
>             'P2': ['chords', 'scales', 'singlenotes', 'techniques'],
>             'P3': ['music']
>         }
>         self.valid_modalities = ['directinput', 'micamp', 'exo', 'ego']
> ```
> 
> make these variables uppercase to denote that they are GLOBAL definitions.

fixed in Commit f248e2b

Madhudorai commented:
> ```
>         # Build expanded_index for slicing. Use ceiling division so the last slice may be incomplete.
>         if self.slice_size:
>             self.expanded_index = []
>             for i, sample_meta in enumerate(self.index):
>                 midi_path = self._get_midi_path(sample_meta)
>                 if os.path.exists(midi_path):
>                     midi_obj = pretty_midi.PrettyMIDI(midi_path)
>                     total_duration = midi_obj.get_end_time()
>                     num_slices = int(np.ceil(total_duration / self.slice_size))
>                     for s in range(num_slices):
>                         start = s * self.slice_size
>                         end = (s + 1) * self.slice_size  # May exceed total_duration; will be padded.
>                         self.expanded_index.append((i, start, end))
>         elif self.slice_range:
>             self.expanded_index = [(i, self.slice_range[0], self.slice_range[1])
>                                    for i in range(len(self.index))]
>         else:
>             # No slicing: entire sample is one slice.
>             self.expanded_index = [(i, None, None) for i in range(len(self.index))]
> ```
> 
> This block of code I suggest you do it using vectorized operations. Look at librosa's audio slicing utilities.

do you mean instead of using a for loop, to calculate each slice's start and end?

iranroman commented:
Thank you @Madhudorai for the changes. This is looking very good!

My main concern right now continues being the slicing of audio files. You should be able to do it using librosa's slicing utilities. Please have a look at those and revise your code. You'll also need to do some checks to make sure that the duration of the audio files and the midi are consistent (they should but it's good to double check as there may be small millisecond differences in file duration between different modalities)

Madhudorai committed:
slice overlap, librosa frame util for slicing

Madhudorai commented:
> ```
> - slice_size (float): If specified (in seconds), samples are segmented into contiguous slices of this length.
>                            The last slice, if shorter than slice_size, will be padded with zeros.
> ```
> 
> Change the name of this variable to `slice_dur`. Also state what the behavior is if `None`. I believe the default should not be `None` though.
> 
> You should also have a variable to denote the overlap between slices. Do not assume that slices wont overlap.

changed variable name to slice_dur. and added slice overlap in Commit [7836cf6](https://github.com/Madhudorai/Guitar-TECHS/commit/7836cf6bb4bceb68de3c0e2b8e0e4e05df81107a)
